pytorch_lightning/trainer/training_loop.py

```diff
@@ -269,6 +269,16 @@ def _check_training_step_output(self, training_step_output):
             if training_step_output.grad_fn is None:
                 # TODO: Find why - RuntimeError: Expected to mark a variable ready only once ...
                 raise MisconfigurationException("In manual optimization, `training_step` should not return a Tensor")
+        elif self.trainer.lightning_module.automatic_optimization:
+            if not any((
+                isinstance(training_step_output, torch.Tensor),
+                (isinstance(training_step_output, Mapping)
+                 and 'loss' in training_step_output), training_step_output is None
+            )):
+                raise MisconfigurationException(
+                    "In automatic optimization, `training_step` must either return a Tensor, "
+                    "a dict with key 'loss' or None (where the step will be skipped)."
+                )
```

Comment:
How does this validation work with distributed training?

eg, I noticed a few examples where the loss is not calculated in `training_step` but is instead computed in `training_step_end` after collating the outputs of `training_step`.

Should we maybe swap L299 and L301?



cc @ananthsub